### ISSUE TITLE
Code formatting jumps around

### ISSUE THREAD

**Issue by sascha-andres** [Microsoft/vscode-go#1096]
- VSCode Version: 1.14.1
- OS Version: 

    Linux HGL-049 4.8.0-58-generic #63~16.04.1-Ubuntu SMP Mon Jun 26 18:08:51 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

    DISTRIB_ID=neon
    DISTRIB_RELEASE=16.04
    DISTRIB_CODENAME=xenial
    DISTRIB_DESCRIPTION="KDE neon User Edition 5.10"

Since the VSCode 1.14.1 Update the formatting of go code in larger files ( file used has 113 lines ) jumps around and slects text which then easily gets overwritten. I made a screen recording as this will probably make it easier to understand my point

[Screen recording](https://youtu.be/Zp6WZi-MZVc)

Reproduces without extension: No

**Comment by fyndor** [comment]
Yea this is super annoying and in my opinion makes this writing go code impossible in VS Code.  One fix I have found is turning off auto save but then you lose a lot of instant gratification of the extension telling you when you are writing improper code etc.  I think this is a really important bug to fix

**Comment by ramya-rao-a** [comment]
@sascha-andres, @fyndor 

Sorry, this issue somehow got lost in the sea of notifications I currently have.

Do you have any of the `editor.formatOn...` settings turned on?

**Comment by sascha-andres** [comment]
My go specific settings:

```
  // go extension
  "go.enableCodeLens": {
      "references": true,
      "runtest": true
  },
	"go.lintTool": "golint",
  "go.lintOnSave": "package",
  "go.useLanguageServer": false,
  "go.buildOnSave": "package",
  "go.vetOnSave": "off",
  "go.buildTags": "",
  "go.buildFlags": [],
  "go.lintFlags": [],
  "go.vetFlags": [],
  "go.coverOnSave": false,
  "go.useCodeSnippetsOnFunctionSuggest": false,
  "go.formatOnSave": true,
  "go.formatTool": "goimports",
  "go.gocodeAutoBuild": false,
```

**Comment by ramya-rao-a** [comment]
@sascha-andres My question was around the format settings for the editor like `editor.formatOnSave`, `editor.formatOnPaste`, editor.formatOnType`. Have you enabled any of these?

**Comment by sascha-andres** [comment]
@ramya-rao-a No I do not have any `editor.format...` set. My `editor. ` settings:

```
  // editor settings
  "editor.fontFamily": "Fira Code",
  "editor.fontSize": 16,
  "editor.fontLigatures": true,
  "editor.tabSize": 2,
  "editor.useTabStops": true,
  "editor.minimap.enabled": true,
```

**Comment by ramya-rao-a** [comment]
@sascha-andres Can you disable all other extensions you have and try again? There have been cases when operations from other extensions interfere with the save cycle and triggering edits -> save -> format again.

**Comment by sascha-andres** [comment]
Done this but nothing has changed

**Comment by ramya-rao-a** [comment]
I have a repro after spending an hour on this. Finally!

The root cause is that `formatOnSave` feature uses the `TextEditorEdit` api to apply the formatting edits on the file.
Edits made this way have this issue of ending up selecting part of the text that the cursor is on.

I have logged an upstream bug for this in VS Code:
https://github.com/Microsoft/vscode/issues/32058

To fix the issue from the Go extension's side, I have updated to use `WorkspaceEdit` api instead.

The fix will be out in the next update.

Thanks for logging the issue and being patient with me :)
